### ISSUE TITLE
Helper types for the component props

### ISSUE THREAD

**Issue by kastermester** [kastermester/relay-compiler-language-typescript#25]
This is just me shoving ideas out of my head, we don’t need this for initial release.

I’d like to improve on the experience on this when writing components, ie. having to import the fragment types and making sure you use the right ones, even when adding new properties seem... like not a great experience to me.

I’ve been thinking about this, and so far I’ve come to the conclusion that creating a d.ts file as part of the plugin run which could declare a bunch of types usable throughout ones entire codebase might be helpful. Ie. if we can create a type `MyComponentFragmentContainerProps<TProps>` (better name needed) which more or less does the opposite of the types we need for the definitelytyped repo, ie takes a type object of non relay props and returns a new type containing all that + the fragment types and then the relay prop. We’d probably need different types for the 3 different kind of containers due to differences in the relay prop.

I’m not entirely sure if this is doable without adding more support in the relay compiler itself or not, but if any is needed it shouldn’t be that much, and I’m sure we could make something flow users can use as well.

Let me know if I need to elaborate (I’m typing this from my iPad atm) the ideas, or if you have different ideas for making this nicer.

**Comment by kastermester** [comment]
Cc: @alloy

**Comment by alloy** [comment]
This sounds like a great idea!

I was thinking about something similar, but different, recently; which is to provide typed aliases of `commitMutation` in the generated artefacts such that you don’t need to do anything else then `import { commitMyModuleMutation } from "__generated__/MyModuleMutation.graphql"`.

**Comment by kastermester** [comment]
The initial thoughts I had for this I've actually since found out is not ideal (from working with our own codebase using some slightly different techniques but with regards to this issue, same stuff applies).

I would however still like to approach the feeling of just being able to add more fragments and have it working without any further code modifications - I'm just not exactly sure how that would work.

How would you want your idea to work? Like could you show some short examples of what would be generated and how you would call it?

**Comment by alloy** [comment]
Currently we generate types for the [input variables and response](https://github.com/artsy/reaction/blob/175eeaf19ff585e346501365b1d24c2198770d51/src/__generated__/ArtistSearchResultsArtistMutation.graphql.ts#L4-L48), but casting `commitMutation` is left up to the user –currently I’m not even sure that `commitMutation` accepts the proper type arguments, but that aside–. I’m thinking that besides generating these types, we also emit something like:

```ts
import { commitMutation, CommitMutation } from "relay-runtime"
export const commitArtistSearchResultsArtistMutation: CommitMutation<ArtistSearchResultsArtistMutationVariables, ArtistSearchResultsArtistMutationResponse> = commitMutation
```

The user can then do:

```ts
import { commitArtistSearchResultsArtistMutation as commitMutation } from "__generated__/ArtistSearchResultsArtistMutation.graphql"
```

Or without the import alias in case a module has multiple mutations.

**Comment by alloy** [comment]
Yeah right now `commitRelayModernMutation` doesn’t take any type parameters for either variables or response, so that needs to be fixed anyways.

**Comment by kastermester** [comment]
Ah I see what you mean. In general I think getting variables and output type needed for a query and mutation (and subscriptions for that matter) texts would be very nice.

Something like:

MyQuery.ts
```
interface MyQueryVariablesAndResult {
  " $variables": VariablesType;
  " $result": ResultType;
}
const node: ConcreteRequest & MyQueryVariablesAndResult = (function(){ ... ])();
```

Then we could type the commitMutation (and stuff around the QueryRenderer) to work with the ` $variables` and ` $result` properties if they exist (using the awesome stuff from TS 2.8 that we made use for for component prop types).

**Comment by alloy** [comment]
I’m not sure I see how that would work, can you provide an example of the `commitMutation` function? The thing is that, unlike the `create*Container` functions, we never pass any type to `commitMutation` from which we could infer any typings, we just pass data.

**Comment by kastermester** [comment]
Something like this:

https://www.typescriptlang.org/play/#src=%2F%2F%20~~~~~~~~~~~~~~~~~~~~~%0D%0A%2F%2F%20commitRelayModernMutation%0D%0A%2F%2F%20~~~~~~~~~~~~~~~~~~~~~%0D%0A%2F%2F%20exposed%20through%20RelayModern%2C%20not%20Runtime%20directly%0D%0Aexport%20interface%20MutationConfig%3CTMutation%20extends%20GraphQLTaggedNode%3E%20%7B%0D%0A%20%20%20%20variables%3A%20OperationVariables%3CTMutation%3E%3B%0D%0A%20%20%20%20mutation%3A%20TMutation%3B%0D%0A%20%20%20%20onCompleted%3F(response%3A%20OperationResult%3CTMutation%3E)%3A%20void%3B%0D%0A%7D%0D%0A%0D%0Aexport%20interface%20Environment%20%7B%20%7D%0D%0Aexport%20interface%20Disposable%20%7B%0D%0A%20%20%20%20dispose()%3A%20void%3B%0D%0A%7D%0D%0A%0D%0Aexport%20type%20ConcreteFragment%20%3D%20any%3B%0D%0A%0D%0Aexport%20type%20GraphQLTaggedNode%20%3D%0D%0A%20%20%20%20%7C%20(()%20%3D%3E%20ConcreteFragment)%0D%0A%20%20%20%20%7C%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20modern()%3A%20ConcreteFragment%3B%0D%0A%20%20%20%20%20%20%7D%3B%0D%0A%0D%0Ainterface%20OperationWithVariables%3CTVars%3E%20%7B%0D%0A%20%20%20%20%22%20%24variables%22%3A%20TVars%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20OperationWithResult%3CTResult%3E%20%7B%0D%0A%20%20%20%20%22%20%24result%22%3A%20TResult%3B%0D%0A%7D%0D%0A%0D%0Aexport%20type%20OperationVariables%3CTMutation%20extends%20GraphQLTaggedNode%3E%20%3D%20TMutation%20extends%20OperationWithVariables%3Cinfer%20V%3E%20%3F%20V%20%3A%20any%3B%0D%0Aexport%20type%20OperationResult%3CTMutation%20extends%20GraphQLTaggedNode%3E%20%3D%20TMutation%20extends%20OperationWithResult%3Cinfer%20R%3E%20%3F%20R%20%3A%20any%3B%20%0D%0A%0D%0Aexport%20declare%20function%20commitRelayModernMutation%3CTMutation%20extends%20GraphQLTaggedNode%3E(environment%3A%20Environment%2C%20config%3A%20MutationConfig%3CTMutation%3E)%3A%20Disposable%3B%0D%0A%0D%0A%2F%2F%20Import%20this%20from%20artifact%20file%0D%0Adeclare%20const%20MyOperation%3A%20GraphQLTaggedNode%20%26%20%7B%0D%0A%20%20%20%20%22%20%24variables%22%3A%20%7B%0D%0A%20%20%20%20%20%20%20%20myVar%3A%20string%3B%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20%22%20%24result%22%3A%20%7B%0D%0A%20%20%20%20%20%20%20%20viewer%3A%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20profile%3A%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20name%3A%20string%3B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20%7D%0D%0A%20%20%20%20%20%20%20%20%7D%0D%0A%20%20%20%20%7D%0D%0A%7D%3B%0D%0A%0D%0Aconst%20myMutation%20%3D%20commitRelayModernMutation(%7B%7D%2C%20%7B%0D%0A%20%20%20%20mutation%3A%20MyOperation%2C%0D%0A%20%20%20%20variables%3A%20%7B%0D%0A%20%20%20%20%20%20%20%20myVar%3A%20'test'%2C%0D%0A%20%20%20%20%7D%2C%0D%0A%20%20%20%20onCompleted(res)%20%7B%0D%0A%20%20%20%20%20%20%20%20console.log(res.viewer.profile.name)%3B%0D%0A%20%20%20%20%7D%0D%0A%7D)

**Comment by alloy** [comment]
Hmm, but the `mutation` parameter is supposed to be the GraphQL document, so it can’t exist in the artefact file.

**Comment by kastermester** [comment]
I think I might have messed up some details in the link, [here's an updated version](https://www.typescriptlang.org/play/#src=%2F%2F%20~~~~~~~~~~~~~~~~~~~~~%0D%0A%2F%2F%20commitRelayModernMutation%0D%0A%2F%2F%20~~~~~~~~~~~~~~~~~~~~~%0D%0A%2F%2F%20exposed%20through%20RelayModern%2C%20not%20Runtime%20directly%0D%0Aexport%20interface%20MutationConfig%3CTMutation%20extends%20ConcreteRequest%3E%20%7B%0D%0A%20%20%20%20variables%3A%20OperationVariables%3CTMutation%3E%3B%0D%0A%20%20%20%20mutation%3A%20TMutation%3B%0D%0A%20%20%20%20onCompleted%3F(response%3A%20OperationResult%3CTMutation%3E)%3A%20void%3B%0D%0A%7D%0D%0A%0D%0Aexport%20interface%20Environment%20%7B%20%7D%0D%0Aexport%20interface%20Disposable%20%7B%0D%0A%20%20%20%20dispose()%3A%20void%3B%0D%0A%7D%0D%0A%0D%0Aexport%20type%20ConcreteFragment%20%3D%20any%3B%0D%0Aexport%20type%20ConcreteRequest%20%3D%20any%3B%0D%0A%0D%0Ainterface%20OperationWithVariables%3CTVars%3E%20%7B%0D%0A%20%20%20%20%22%20%24variables%22%3A%20TVars%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20OperationWithResult%3CTResult%3E%20%7B%0D%0A%20%20%20%20%22%20%24result%22%3A%20TResult%3B%0D%0A%7D%0D%0A%0D%0Aexport%20type%20OperationVariables%3CTMutation%20extends%20ConcreteRequest%3E%20%3D%20TMutation%20extends%20OperationWithVariables%3Cinfer%20V%3E%20%3F%20V%20%3A%20any%3B%0D%0Aexport%20type%20OperationResult%3CTMutation%20extends%20ConcreteRequest%3E%20%3D%20TMutation%20extends%20OperationWithResult%3Cinfer%20R%3E%20%3F%20R%20%3A%20any%3B%20%0D%0A%0D%0Aexport%20declare%20function%20commitRelayModernMutation%3CTMutation%20extends%20ConcreteRequest%3E(environment%3A%20Environment%2C%20config%3A%20MutationConfig%3CTMutation%3E)%3A%20Disposable%3B%0D%0A%0D%0A%2F%2F%20Import%20this%20from%20artifact%20file%0D%0Aconst%20MyOperation%3A%20ConcreteRequest%20%26%20%7B%0D%0A%20%20%20%20%22%20%24variables%22%3A%20%7B%0D%0A%20%20%20%20%20%20%20%20myVar%3A%20string%3B%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20%22%20%24result%22%3A%20%7B%0D%0A%20%20%20%20%20%20%20%20viewer%3A%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20profile%3A%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20name%3A%20string%3B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20%7D%0D%0A%20%20%20%20%20%20%20%20%7D%0D%0A%20%20%20%20%7D%0D%0A%7D%20%3D%20(function%20()%20%7B%0D%0A%20%20%20%20%20%20%20%20var%20v0%20%3D%20%5B%2F*...*%2F%5D%3B%0D%0A%20%20%20%20%20%20%20%20var%20v1%20%3D%20%5B%2F*...*%2F%5D%3B%0D%0A%20%20%20%20%20%20%20%20return%20%7B%20%2F*%20some%20object%20representing%20the%20actual%20mutation%20*%2F%20%7D%3B%0D%0A%7D)()%20as%20any%3B%0D%0A%0D%0Aconst%20myMutation%20%3D%20commitRelayModernMutation(%7B%7D%2C%20%7B%0D%0A%20%20%20%20mutation%3A%20MyOperation%2C%0D%0A%20%20%20%20variables%3A%20%7B%0D%0A%20%20%20%20%20%20%20%20myVar%3A%20'test'%2C%0D%0A%20%20%20%20%7D%2C%0D%0A%20%20%20%20onCompleted(res)%20%7B%0D%0A%20%20%20%20%20%20%20%20console.log(res.viewer.profile.name)%3B%0D%0A%20%20%20%20%7D%0D%0A%7D)) that might show what I'm after.

In short, the actual GraphQL document we export out of the artifact file I want to decorate it with extra typing information regarding it's variables and it's result type (ie. the payload of the mutation/subscription/query). This way we can add generic type parameters and infer the variables and result based off the operation itself (without providing extra generic type parameters) as shown in the example.

**Comment by alloy** [comment]
Are you thinking of the `MyOperation` query being replaced by e.g. the babel plugin for the processed operation? Because that doesn’t help at dev time when it’s just supposed to be a tagged template string.

**Comment by kastermester** [comment]
I think I know what you're getting at now. I've completely forgotten how this is supposed to work exactly, my bad.

With that in mind, you're right we cannot make infer the variable and payload types without a generic type argument. Perhaps that is alright though? I mean there's not much difference to calling:

```ts
commitMutation<AddProfileMutation>(...);
```

As opposed to:

```ts
commitAddProfileMutation(...);
```

If this is not desirable enough then we either need to look into doing something like you described - or see if there's some TypeScript specific extensions we can utilize to get the types flowing based off the template literals.

**Comment by kastermester** [comment]
To build on the idea of creating a TypeScript plugin it should be possible to add a [Language Server Plugin](https://github.com/Microsoft/TypeScript/wiki/Writing-a-Language-Service-Plugin) that would add linting quick fixes to add the generic type argument (and import it from the proper file). How does this sound to you?

**Comment by alloy** [comment]
Ooooh, that sounds amazing 👌

But yeah, I agree, just adding the type argument should suffic. I guess I was reasoning based off how we currently emit 2 types for variables and response and specifying those both each time feels a little tedious.... but we obviously control how that gets emitted and so should just fix it there 😄

**Comment by alloy** [comment]
The `commitMutation` (and `QueryRenderer`) part is done.

As to your OP, I think we need to expand the plugin interface and have relay-compiler specify the type of container so we can appropriately type the `relay` parameter. Also, I’m thinking that we can just add to the existing artefact files and without the need for a factory type. E.g. for a refetch container we could emit something like:

```ts
interface MyRefetch_foo_props {
  relay: RelayRefetchProp
  foo: MyRefetch_foo
}
```

Which is then used in the user’s code extend their props interface from.

(I’m not a big fan of the interface naming with `_props`, but coming up short with ways to improve that)
